src/main/plugin/iso19139.ca.HNAP/layout/tpl-csv.xsl

```diff
@@ -0,0 +1,211 @@
+      </metadatacreationdate>
+      <referencesystem>
+        <xsl:value-of select="gmd:referenceSystemInfo/gmd:MD_ReferenceSystem/gmd:referenceSystemIdentifier/gmd:RS_Identifier/gmd:code/gco:CharacterString"/>
+      </referencesystem>
+
+      <!-- For each date grouped by the date type mapped to a readable value -->
+      <xsl:for-each select="gmd:identificationInfo/*/gmd:citation/*/gmd:date">
+        <xsl:variable name="dateTypeCode" select="*/gmd:dateType/*/@codeListValue"/>
+        <xsl:variable name="dateTypeCodeReadable" select="tokenize($codelists/codelist[@name = 'gmd:CI_DateTypeCode']/entry[code/text() = $dateTypeCode]/value/text(), ';')[1]"/>
+
+        <xsl:element name="date-{$dateTypeCodeReadable}">
+          <xsl:value-of select="*/gmd:date/*/text()"/>
+        </xsl:element>
+      </xsl:for-each>
+
+      <xsl:for-each select="gmd:identificationInfo/*/gmd:graphicOverview/*/gmd:fileName">
+        <image>
+          <xsl:value-of select="*/text()"/>
+        </image>
+      </xsl:for-each>
+
+
+      <!-- All keywords not having thesaurus reference or an empty thesaurusName -->
+      <xsl:for-each select="gmd:identificationInfo/*/gmd:descriptiveKeywords/gmd:MD_Keywords[not(gmd:thesaurusName) or gmd:thesaurusName/*/gmd:title/(gco:CharacterString|gmx:Anchor)/text() = '']">
+        <xsl:variable name="keywordTypeCode" select="gmd:type/*/@codeListValue"/>
+        <xsl:variable name="keywordTypeCodeReadable" select="tokenize($codelists/codelist[@name = 'gmd:MD_KeywordTypeCode']/entry[code/text() = $keywordTypeCode]/value/text(), ';')[1]"/>
+
+        <xsl:for-each select="gmd:keyword[not(@gco:nilReason) or */text() != '']">
+          <xsl:choose>
+            <!-- All keywords with a valid keywordTypeCode -->
+            <xsl:when test="$keywordTypeCodeReadable != ''">
+              <xsl:element name="keyword-{$keywordTypeCodeReadable}">
+                <xsl:apply-templates mode="localised" select=".">
+                  <xsl:with-param name="langId" select="$langId"/>
+                </xsl:apply-templates>
+              </xsl:element>
+            </xsl:when>
+
+            <!-- All keywords without a valid keywordTypeCode -->
+            <xsl:otherwise>
+              <keyword-other>
+                <xsl:apply-templates mode="localised" select=".">
+                  <xsl:with-param name="langId" select="$langId"/>
+                </xsl:apply-templates>
+              </keyword-other>
+            </xsl:otherwise>
+          </xsl:choose>
+        </xsl:for-each>
+      </xsl:for-each>
+
+      <!-- All keywords with a valid thesaurus name -->
+      <xsl:for-each select="gmd:identificationInfo/*/gmd:descriptiveKeywords/gmd:MD_Keywords[gmd:thesaurusName/*/gmd:title/(gco:CharacterString|gmx:Anchor)/text() != '']">
+        <xsl:variable name="thesaurusName">
+          <xsl:variable name="translationLanguage" select="gmd:thesaurusName/*/gmd:title/*/*/gmd:LocalisedCharacterString/@locale"/>
```

Comment:
gmd:LocalisedCharacterString/@locale is incorrect as it is a local code/id and not the actual languagues.

If you need to get the language then you should use the value withing `$langId` which identifies the main language of the document - There are other xsl that collect the main language and alt language and you can refer to that code.

Code should never check for the id. i.e. `#eng` is just and id and it could be used to identify French language documents (although highly unlikely).  A more accurate example could be French being represented with id #fra or #fre.   In either case if you really need the language then you have to map the id #eng to the official language code to ensure the id is the correct language.   There is multiple xslt that does this so I suggest that you look at other xsl for samples.

Reply:
https://github.com/metadata101/iso19139.ca.HNAP/blob/859ede0bc2648332cd1044a0e114d57e11d6f2f4/src/main/plugin/iso19139.ca.HNAP/layout/tpl-csv.xsl#L121-L128

@ianwallen Is this what you had in mind regarding the localization of thesaurus names? I am now getting the locale id that matches the language code for English and using it within the `localised` template to get the correct translation.

I am working under the assumption that the csv export column headers should always be in English for the consistency of exported data. Let me know if all of the element names should be localized as well.

Reply:
Testing results from code from this branch:

- English UI - search result export with English main language -> English column headers, English content
- English UI - search result export with French main language -> English column headers, English content
- French UI - search result export with English main language -> English column headers, French content
- French UI - search result export with French main language -> English column headers, French content (except for mapped category value)

Testing results from prod:

- English UI - search result export with English main language -> English column headers, English content
- English UI - search result export with French main language -> English column headers, English content
- French UI - search result export with English main language -> English column headers, Mixed content (fra remains in English, fre is in French)
- French UI - search result export with French main language -> English column headers, French content (except for mapped category value)